libraries/ms-helper/src/mcp2515.c

```diff
@@ -0,0 +1,235 @@
+  uint8_t payload[] = { MCP2515_CMD_RESET };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), NULL, 0);
+}
+
+static void prv_read(Mcp2515Storage *storage, uint8_t addr, uint8_t *read_data, size_t read_len) {
+  uint8_t payload[] = { MCP2515_CMD_READ, addr };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), read_data, read_len);
+}
+
+static void prv_write(Mcp2515Storage *storage, uint8_t addr, uint8_t *write_data,
+                      size_t write_len) {
+  uint8_t payload[MCP2515_MAX_WRITE_BUFFER_LEN];
+  payload[0] = MCP2515_CMD_WRITE;
+  payload[1] = addr;
+  memcpy(&payload[2], write_data, write_len);
+  spi_exchange(storage->spi_port, payload, sizeof(payload), write_data, write_len);
+}
+
+static void prv_bit_modify(Mcp2515Storage *storage, uint8_t addr, uint8_t mask, uint8_t data) {
+  uint8_t payload[] = { MCP2515_CMD_BIT_MODIFY, addr, mask, data };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), NULL, 0);
+}
+
+static uint8_t prv_read_status(Mcp2515Storage *storage) {
+  uint8_t payload[] = { MCP2515_CMD_READ_STATUS };
+  uint8_t read_data[1] = { 0 };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), read_data, sizeof(read_data));
+
+  return read_data[0];
+}
+
+static void prv_handle_rx(Mcp2515Storage *storage, uint8_t int_flags) {
+  for (size_t i = 0; i < SIZEOF_ARRAY(s_rx_buffers); i++) {
+    Mcp2515RxBuffer *rx_buf = &s_rx_buffers[i];
+    if (int_flags & rx_buf->int_flag) {
+      // message RX
+
+      // Read ID
+      uint8_t id_payload[] = { MCP2515_CMD_READ_RX | rx_buf->id };
+      uint8_t read_id[5] = { 0 };
+      spi_exchange(storage->spi_port, id_payload, sizeof(id_payload), read_id, sizeof(read_id));
+
+      // Unpack ID: SIDH, SIDL, EID8, EID0, RTSnDLC
+      //
+      // STD[10:3] in SIDH[7:0], STD[2:0] in SIDL[7:5]
+      // EXT[17:16] in SIDL[1:0], EXT[15:8] in EID8[15:8], EXT[7:0] in EID0[7:0]
+      // extended bit in SIDL[3]
+      // TODO(ELEC-462): don't use magic numbers?
+      uint32_t id = (uint32_t)((read_id[0] << 3) & 0xF) | (uint32_t)((read_id[1] >> 5) & 0x3) |
+                    (uint32_t)((read_id[1] << 27) & 0x2) | (uint32_t)(read_id[2] << 19) |
+                    (uint32_t)(read_id[3] << 11);
+      bool extended = (read_id[1] >> 3) & 0x1;
+
+      if (!extended) {
+        // Standard IDs have garbage in the extended fields
+        id &= 0x7FF;
+      }
+      size_t dlc = read_id[4] & MCP2515_TXBNDLC_DLC_MASK;
+
+      uint8_t data_payload[] = { MCP2515_CMD_READ_RX | rx_buf->data };
+      uint64_t read_data = 0;
+      spi_exchange(storage->spi_port, data_payload, sizeof(data_payload), (uint8_t *)&read_data,
+                   sizeof(read_data));
+
+      if (storage->rx_cb != NULL) {
+        storage->rx_cb(id, extended, read_data, dlc, storage->context);
+      }
+    }
+  }
+}
+
+static void prv_handle_error(Mcp2515Storage *storage, uint8_t int_flags) {
+  if (int_flags & MCP2515_CANINT_EFLAG) {
+    // Do we bother recording errors?
```

Comment:
Debatable and plumbing any kind of error out of here is probably non-trivial. I would recommend a `LOG_DEBUG` or just leave the comment but make it not a question :p